In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

path_to_data = "data_pizdata"
random_state = 42

In [32]:
# В СКАЧАНЫХ ДАНЫХ НЕТ НАЗВАНИЯ СТОЛБЦОВ и он берёт за названия первую строку

df_x_train = pd.read_csv(os.path.join(path_to_data, "X_train.csv"), delimiter=',', header=None)
df_y_train = pd.read_csv(os.path.join(path_to_data, "y_train.csv"), delimiter=',', header=None)
df_test = pd.read_csv(os.path.join(path_to_data, "X_test.csv"), delimiter=',', header=None)

df_x_train.columns = [i for i in range(df_x_train.shape[1])]
df_test.columns = df_x_train.columns

In [33]:
df_x_train.head()

0         1         2         3    4    5         6         7   \
0  0.451009  0.614606  0.167306  0.390843  0.0  2.0  6.691263  6.792145   
1  0.837123  0.558449  0.034198  0.839781  1.0  2.0  0.735178  0.324811   
2  0.605636  0.350955  0.678408  0.568693  1.0  0.0  3.921524  6.610859   
3  0.862016  0.205944  0.576584  0.198756  0.0  2.0  1.873797  4.482543   
4  0.586450  0.767571  0.751014  0.642278  1.0  0.0  0.926542  6.928612   

         8         9         10        11   12   13        14        15  
0  1.185154  3.811872  0.175758  2.207100  0.0  2.0  6.691263  6.792145  
1  1.575394  5.370452  6.851491  2.479178  1.0  2.0  0.735178  0.324811  
2  1.959624  0.698998  6.002673  0.780729  1.0  0.0  3.921524  6.610859  
3  1.652851  5.640393  0.702110  2.102840  0.0  2.0  1.873797  4.482543  
4  1.676833  5.206752  4.588336  2.963444  1.0  0.0  0.926542  6.928612

In [34]:
df_test.head()

0         1         2         3    4    5    6         7         8   \
0  0.048373  0.466911  0.389070  0.321250  0.0  0.0  0.0  0.000000  0.279088   
1  0.867465  0.621720  0.343289  0.186093  1.0  0.0  0.0  0.000000  1.336795   
2  0.814926  0.433053  0.164681  0.977219  0.0  0.0  0.0  0.000000  1.593938   
3  0.432321  0.425449  0.191574  0.026475  0.0  0.0  0.0  0.000000  0.012634   
4  0.939396  0.433556  0.471259  0.727408  0.0  0.0  0.0  6.193863  1.504852   

         9         10        11   12   13        14        15  
0  3.198103  5.233288  1.622723  1.0  1.0  0.632456  4.717549  
1  1.227167  7.584822  1.235540  1.0  0.0  2.343988  6.778029  
2  5.118116  3.151805  1.382042  1.0  0.0  2.673265  0.511859  
3  2.085054  1.525353  0.920855  1.0  0.0  4.940645  3.900637  
4  4.992094  5.094022  1.900915  1.0  0.0  0.917297  6.193863

In [35]:
df_x_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       40000 non-null  float64
 1   1       40000 non-null  float64
 2   2       40000 non-null  float64
 3   3       40000 non-null  float64
 4   4       40000 non-null  float64
 5   5       40000 non-null  float64
 6   6       40000 non-null  float64
 7   7       40000 non-null  float64
 8   8       40000 non-null  float64
 9   9       40000 non-null  float64
 10  10      40000 non-null  float64
 11  11      40000 non-null  float64
 12  12      40000 non-null  float64
 13  13      40000 non-null  float64
 14  14      40000 non-null  float64
 15  15      40000 non-null  float64
dtypes: float64(16)
memory usage: 4.9 MB


In [36]:
df_x_train.describe()

0             1             2             3             4   \
count  40000.000000  40000.000000  40000.000000  40000.000000  40000.000000   
mean       0.500471      0.498729      0.502726      0.499792      0.499500   
std        0.289359      0.288434      0.289009      0.290115      0.500006   
min        0.000018      0.000020      0.000011      0.000021      0.000000   
25%        0.251133      0.248532      0.252921      0.248177      0.000000   
50%        0.499611      0.499261      0.504818      0.497016      0.000000   
75%        0.751907      0.747167      0.753530      0.753823      1.000000   
max        0.999999      0.999998      0.999978      0.999991      1.000000   

                 5             6             7             8             9   \
count  40000.000000  40000.000000  40000.000000  40000.000000  40000.000000   
mean       0.991275      3.492389      3.487816      1.002620      2.994680   
std        0.818209      2.016651      2.018964      0.574658      1.726063   
min        0.000000      0.000220      0.000100      0.000068      0.000015   
25%        0.000000      1.731510      1.733274      0.507572      1.499907   
50%        1.000000      3.502283      3.489752      1.004894      2.992040   
75%        2.000000      5.230650      5.243342      1.500862      4.475840   
max        2.000000      6.999947      6.999826      1.999868      5.999870   

                 10            11            12            13            14  \
count  40000.000000  40000.000000  40000.000000  40000.000000  40000.000000   
mean       4.500824      1.501175      0.499500      0.991275      3.492389   
std        2.594021      0.870372      0.500006      0.818209      2.016651   
min        0.000172      0.000113      0.000000      0.000000      0.000220   
25%        2.271650      0.743164      0.000000      0.000000      1.731510   
50%        4.492366      1.494108      0.000000      1.000000      3.502283   
75%        6.740416      2.260100      1.000000      2.000000      5.230650   
max        8.998587      2.999977      1.000000      2.000000      6.999947   

                 15  
count  40000.000000  
mean       3.487816  
std        2.018964  
min        0.000100  
25%        1.733274  
50%        3.489752  
75%        5.243342  
max        6.999826

In [37]:
import pandas as pd
from tqdm import tqdm
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from catboost import CatBoostRegressor
from sklearn.metrics import mean_squared_error

X_train, X_test, y_train = df_x_train, df_test, df_y_train

# Взял лучшие модели, просто посмотреть какие лучше решают задачу, в базовой комплектации
models = {
    'RandomForest': RandomForestRegressor(random_state=random_state),
    'GradientBoosting': GradientBoostingRegressor(random_state=random_state),
    'CatBoost': CatBoostRegressor(random_state=random_state, verbose=0)  # Это тоже градиентный бустинг тока от катбуста
}

param_grids = {
    'RandomForest': {
        'n_estimators': [100, 200, 300],
        'max_depth': [None, 10, 20, 30],
        'min_samples_split': [2, 5, 10]
    },
    'GradientBoosting': {
        'n_estimators': [100, 200, 300],
        'learning_rate': [0.01, 0.1, 0.2],
        'max_depth': [3, 5, 7]
    },
    'CatBoost': {
        'iterations': [100, 200, 300],
        'learning_rate': [0.01, 0.1, 0.2],
        'depth': [4, 6, 8]
    }
}

best_models = {}
best_scores = {}

# Перебираем в тупую все гиперпараметры и смотрим на какиз лос минимален для каждой модели
for name, model in tqdm(models.items(), desc="Подбор гирпараметров", unit="модель"):
    print(f"\nПодбор гиперпараметров для {name}...")
    
    search = RandomizedSearchCV(
        estimator=model,
        param_distributions=param_grids[name],
        n_iter=10,
        scoring='neg_mean_squared_error',  
        cv=3,
        random_state=42,
        n_jobs=-1,
        verbose=1  
    )
    
    
    search.fit(X_train, y_train)
    
    best_models[name] = search.best_estimator_
    best_scores[name] = -search.best_score_
    print(f"Лучший результат для {name} (MSE): {best_scores[name]:.4f}")

# Выбираем по минимальному лосу лучшую модель
best_model_name = min(best_scores, key=best_scores.get)
best_model = best_models[best_model_name]
print(f"\nЛучшая модель: {best_model_name} с MSE: {best_scores[best_model_name]:.4f}")


print("Выполняем предсказания...")
y_pred = best_model.predict(X_test)

sumbission = pd.DataFrame(y_pred, columns=['y'])
sumbission['ID'] = list(range(len(y_pred)))
sumbission.to_csv('predictions.csv', sep=',', index=False)
print("Предсказания сохранены в файл predictions.csv")


Подбор гиперпараметров:   0%|          | 0/3 [00:00<?, ?модель/s]


Подбор гиперпараметров для RandomForest...



Обучение RandomForest:   0%|          | 0/10 [00:00<?, ?итерация/s]

Fitting 3 folds for each of 10 candidates, totalling 30 fits


D:\Document\SecCourseMATMEX\ml\FirstKaggleCompetition\venv\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)

Подбор гиперпараметров:  33%|███▎      | 1/3 [04:56<09:52, 296.21s/модель]

Лучший результат для RandomForest (MSE): 3.4659

Подбор гиперпараметров для GradientBoosting...



Обучение GradientBoosting:   0%|          | 0/10 [00:00<?, ?итерация/s]

Fitting 3 folds for each of 10 candidates, totalling 30 fits


D:\Document\SecCourseMATMEX\ml\FirstKaggleCompetition\venv\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?

Подбор гиперпараметров:  67%|██████▋   | 2/3 [10:13<05:08, 308.64s/модель]

Лучший результат для GradientBoosting (MSE): 0.8232

Подбор гиперпараметров для CatBoost...



Обучение CatBoost:   0%|          | 0/10 [00:00<?, ?итерация/s]

Fitting 3 folds for each of 10 candidates, totalling 30 fits



Подбор гиперпараметров: 100%|██████████| 3/3 [10:41<00:00, 213.87s/модель]

Лучший результат для CatBoost (MSE): 0.1582

Лучшая модель: CatBoost с MSE: 0.1582
Выполняем предсказания...
Предсказания сохранены в файл predictions.csv


### Результаты сравнения алгоритмов

Лучший результат для RandomForest (MSE): 3.4659
Лучший результат для GradientBoosting (MSE): 0.8232
Лучшая модель: CatBoost с MSE: 0.1582


In [ ]:
#TODO добавить валидацию
#TODO убрать выбросы и пропуски в данных
#TODO посмотреть ноутбуки